backend/associations/views/election.py

```diff
@@ -1,92 +1,220 @@
-        * Update an election:           PATCH   /associations/elections/1/
-        * Destroy an election:          DELETE  /associations/elections/1/
-        * Vote to an election:          POST    /associations/elections/1/vote/
-        * Results of an election:       GET     /associations/elections/1/results/
-"""
+        serializer.save()
+
+    def perform_update(self, serializer):
+        choice = self.get_object()
+        override = {"election": choice.election}
+
+        # If the election has started, do not change the name.
+        if self.get_object().election.started:
+            override["name"] = choice.name
+
+        serializer.save(**override)
+
+    def perform_destroy(self, instance: Choice):
+        # Only allow deletion if the election has not started.
+        if instance.election.started:
+            raise ValidationError("The choice cannot be deleted anymore.")
+
+        instance.delete()
 
 
 class ElectionViewSet(viewsets.ModelViewSet):
     queryset = Election.objects.all()
-    serializer_class = ElectionSerializer
+    serializer_class = ElectionReadOnlySerializer
     permission_classes = (ElectionPermission,)
 
-    @action(detail=True, methods=("get",), permission_classes=(ResultsPermission,))
-    def results(self, *args, **kwargs):
-        election = self.get_object()
-        data = {"election": election.id, "results": election.results}
-        return Response(data=data, status=status.HTTP_200_OK)
+    def get_serializer_class(self):
+        if self.action == "create":
+            return ElectionSerializer
+        elif self.action == "list":
+            return ElectionReadOnlySerializer
 
+        role = self.request.user.get_role(self.get_object().association)
 
-class CreateBallotView(generics.CreateAPIView):
-    queryset = Ballot.objects.all()
-    serializer_class = BallotSerializer
-    permission_classes = (BallotPermission,)
+        if role and role.election:
+            return ElectionSerializer
 
-    def get_election_or_404(self, **kwargs):
-        request_election = Election.objects.filter(pk=kwargs.get("election_pk"))
+        return ElectionReadOnlySerializer
 
-        if not request_election.exists():
-            raise NotFound("Election not found.")
+    def perform_update(self, serializer):
+        # If the election has started, only allow to change `results_are_published`.
+        election = self.get_object()
 
-        return request_election[0]
+        override = (
+            {
+                key: getattr(election, key)
+                for key in ("name", "starts_at", "ends_at", "max_choices_per_ballot")
+            }
+            if election.started
+            else {}
+        )
 
-    def create(self, request, *args, **kwargs):
-        election = self.get_election_or_404(**kwargs)
+        serializer.save(
+            association=election.association,
+            choices=election.choices,
+            voters=election.voters,
+            **override
```

Comment:
Why do we need to override ?

Reply:
These fields will be let alone be  default, won't they ?

Reply:
The serializer currently used does pop these fields during the create, but it's no harm enforcing this rule here too

Reply:
Yes… I really don't understand what you want to change here sorry

`override` is either `{}` or `{ name: ..., ... }` depending on `election.started`